### ISSUE TITLE
would the community be willing to accept an AID export feature?

### ISSUE THREAD

**Issue by dhh1128** [WebOfTrust/signify-ts#274]
I am interested in giving Signify an AID export feature (not a whole wallet export feature), where you could ask Signify to look up a particular single-sig AID's current priv+pub keypair and its next public key (the one it's committed to in prerotation), and give them back as raw key values.

Before you say it, I KNOW that giving back private keys is dangerous. It would obviously expose the secrets that are the basis of a given AID (except for the next private key) -- so if the person receiving these secrets manages them in a way that is less secure than Signify does, the trust in the AID could be lowered as a result. We have wallets for a reason...

I also know that if Signify is using keys from an HSM, what I've proposed is impossible; you can't get the keys back at all. You can only get a signature from it. So there would be cases where the feature wouldn't work.

So why do I want this? Well, because if I have the keys, I can build bridges to places that know nothing about KERI (yet), but that do understand Ed25519 keys. There are lots of places like that. For example, I can make an SSH key that I use to log in to servers or my github account. I can sign stuff with ssh-keygen. I can get an X509 with the public key from my AID. I can also make a did:key and did:peer (and other DID types too, I suppose) from the AID. And I can make other DID types from it, too. And all of those new places where I'm using my key are deriving their security from stuff that's still held in my Signify wallet, so I can think of the AID as being a root of trust for a whole family of derivative identifiers. The KERI ecosystem now has the option of telling people it has support for SSH, for X509, etc. I can also back up the key material in a password manager.

We could possibly reduce the risk of this feature by only exposing it when an AID is created or its keys are rotated, not when it is in a steady state. And we could expose it in a UI only after a consent screen that explains the risks.

If the community is willing to accept such a feature, I would be willing to invest the time to build it. But I don't want to work on it if it will be rejected for security reasons. So I thought I'd ask the question first. What do you think of the general idea?

Tagging @m00sey , @pfeairheller , @smithsamuelm, @rodolfomiranda

**Comment by SmithSamuelM** [comment]
So why do I want this? Well, because if I have the keys, I can build bridges to places that know nothing about KERI (yet), but that do understand Ed25519 keys. There are lots of places like that. For example, I can make an SSH key that I use to log in to servers or my github account. I can sign stuff with ssh-keygen. I  My first reaction is why not just delegate the other keys as the bridge. The other places don’t understand KERI, so they don’t care about the delegation, but anyone who understands KERI can be presented with the delegation of the “other” keys and then verifiy their relattionship to the Signify keys.This way you never violate the principle of don’t move private keys (secrets) between infrastructures. You always create private keys in the infrastructure where they are to be used (ssh etc) and after the fact just bless them with a delegation from your Signify AID.  Those delegated keys can be effectively ephemeral by making them non-rotatable in their inception event. The delegated AID can be created after the fact of creating the Ed25519 keys or whateever key pair generation (ECDSAk1,r1).So someone that cares about KERI can verify via the delegated key event that the keysare blessed by the signify AID.  But the bridged non keri aware application doesn’t care who blesses the keys.YOu can also bless the keys with an ACDC albeit less securely than delegation.And you can do both.  For example you can have a service that is KERI aware that accepts presentations of either ACDC or a delegated public key in order to add it to an SSH server plublic key store of allowed clients. Then a client can login to the SSH server with the private key. The client is just a non-KERI aware client, the blessing happens OOB to the SSH protocol itself.Sent from my iPadOn Aug 30, 2024, at 02:10, Daniel Hardman ***@***.***> wrote:﻿
I am interested in giving Signify an AID export feature (not a whole wallet export feature), where you could ask Signify to look up a particular single-sig AID's current priv+pub keypair and its next public key (the one it's committed to in prerotation), and give them back as raw key values.
Before you say it, I KNOW that giving back private keys is dangerous. It would obviously expose the secrets that are the basis of a given AID (except for the next private key) -- so if the person receiving these secrets manages them in a way that is less secure than Signify does, the trust in the AID could be lowered as a result. We have wallets for a reason...
I also know that if Signify is using keys from an HSM, what I've proposed is impossible; you can't get the keys back at all. You can only get a signature from it. So there would be cases where the feature wouldn't work.
So why do I want this? Well, because if I have the keys, I can build bridges to places that know nothing about KERI (yet), but that do understand Ed25519 keys. There are lots of places like that. For example, I can make an SSH key that I use to log in to servers or my github account. I can sign stuff with ssh-keygen. I can get an X509 with the public key from my AID. I can also make a did:key and did:peer (and other DID types too, I suppose) from the AID. And I can make other DID types from it, too. And all of those new places where I'm using my key are deriving their security from stuff that's still held in my Signify wallet, so I can think of the AID as being a root of trust for a whole family of derivative identifiers. The KERI ecosystem now has the option of telling people it has support for SSH, for X509, etc. I can also back up the key material in a password manager.
We could possibly reduce the risk of this feature by only exposing it when an AID is created or its keys are rotated, not when it is in a steady state. And we could expose it in a UI only after a consent screen that explains the risks.
If the community is willing to accept such a feature, I would be willing to invest the time to build it. But I don't want to work on it if it will be rejected for security reasons. So I thought I'd ask the question first. What do you think of the general idea?
Tagging @m00sey , @pfeairheller , @SmithSamuelM, @rodolfomiranda

**Comment by edeykholt** [comment]
Two questions come to mind, @dhh1128:
1) What about a feature where a signify client provides a non-rotatable AID (essentially the public key corresponding to a private key it already has), then delegates to it? Would that feature enable many of your desired use cases?
2) Without a AID private key export feature via Signify, how could a KERIA account owner move their AIDs controlled by one KERIA Service to another provider or platform?  Could they request an export of their entire database, then import it to another KERIA instance, or decrypt its content with tools to extract key material needed?  Without these types of features, is an AID essentially locked in with the KERIA provider?

**Comment by dhh1128** [comment]
@SmithSamuelM said:

>My first reaction is why not just delegate the other keys as the bridge. 

This sounds cool, but I am not aware of a way to delegate pure keys. AFAIK, the only way to delegate in KERI is cooperative delegation, which requires an AID on the receiving/delegatee side. Is there a way to inject an existing key pair into the routine that generates a delegated AID? If so, it could well address most of my intent, and I'd love to learn about it!

@edeykholt said:

>What about a feature

Yes, this is pretty much what I need. Except I don't want a Signify client to provide it. I want to bring the pub key to KERI and have KERI delegate to it. But I don't know a way to do that.

>how could a KERIA account owner move their AIDs... locked in with the KERIA provider?

I think we can write a feature that moves a wallet from one KERIA instance to another. But not a feature that moves individual AIDs.

**Comment by SmithSamuelM** [comment]
This functionality of moving keys around is not really part of KERI. It just so happens that keripy includes a key store and key store manager in the keri.app.keeping module. 

The Manager class has a method called .replay and one called .ingest that will allow one to export i.e.  replay the key history for a given AID and then import i.e ingest a key history created someplace else. 

the Hab.make method already optionally allows one to ingest a key history into its associated key manager and then replay the key at the current key state.

There is a lot of functionality in the keeping module that very few know about because its not part of the KERI protocol but was created to enable basic wallet functionality in keripy.

**Comment by SmithSamuelM** [comment]
@dhh1128   The key manager keri.app.keeping.Manager can ingest private keys created elsewhere. So after the fact you can spin up a Manager and a Hab which Hab looks up the public key from the Manger so that the Hab tha can create the incept message that generates the AID for any set of private keys the Manager has already ingested. So you don't need to know the AID of the delegate until after the fact. So you can decide to delegate after the fact and create the delegated AID on the fly given you already have the key pairs to ingest.   The point of the delegation is to enable some user to authentically attribute that the private keys that after the fact generated the delegated aid, were indeed blessed.  So you can build the bridge afterwards you don't have to spin up a wallet to use the private keys. You can at any time spin up a wallet in the infrastructure that created the key pairs, ingest the key pairs, and then cooperatively delegate them via the resultant aid.

**Comment by SmithSamuelM** [comment]
So for example I can take my ledger nano create keys pairs use them to sign up etc and then later export the public keys to create an aid sign it with the ledger nano, publish it and then have it cooperatively delegated all after the fact of using those private keys via the nano. The nano never needs to know it is being used for KERI. So this is a bridge that backfills when an only when you need the backfill.

**Comment by SmithSamuelM** [comment]
The typical use case for the Keripy keeping.Manager is to NOT create and store private keys directly but to use a salt and an HKKey algorthm to generate the private keys on the fly from a public HDKey path. So moving private keys between infrastructures never happens. One moves the salt out-of-band  and the public HKKey path in band. Moving the salt is easily accomplished by using something like 1password. So we really don't in general need to export anything other than the HDKey path, which actually for an existing KEL can be generated from the KEL itself (just the sn and rotation index of the public key history).   

So if you are generating and storing private keys someplace not using an HDKey mechanism or are using an HSM that doesn't allow you to export the salt. then you basically have private key material you can import after the fact. So it never ever needs to be exported from a KERI wallet only imported to a KERI wallet.

**Comment by SmithSamuelM** [comment]
By the way, using KELs that contribute to a distributed multisig takes advantage of the fact that the key state of the contributing keys can be regenerated from the salt and HDKey path derived from the contributing KEL. So, the key manager doesn't need a different construct to support HDKeys for distributed multisig groups.  If a given controller contributes keys from one of its own KELs then all the key manager has to track is its contributing KEL and the associated salt.

**Comment by SmithSamuelM** [comment]
So back to the original question:  I assumed that if there is an application outside of KERI that needs private key material, then that application already has a way to generate that private key material. So KERI just needs to import it and bless it via delegation.

If that outside application does not already have a way to generate private key material then I strongly suggest that one use an HDKey generation algorithm with a secure Salt. The Salt can be created, stored, and moved between infrastructure using 1password, so no need to reinvent that wheel. So whats left to do. Implement an HDkey generation algorithm externally and track the HDKey path.  If you want to use a KERI wallet for that, you can do that using the keeping.Manager with a non digest temporary AID that is just the fully qualified first public key. Ie. the aid is the first public key. You can move that key history to some other aid later using keeping.Manager.move

Now you can just track the keys and all you have is a placeholder aid. No KEL.  At some time later you can create a real AID and KEL and assign (move) that key history in the the manager to that AID. It creates a forward index so one can track where it got moved to given the original placeholder aid.

Now you don't ever have to export private keys, just move the salt to some local application that can run the HDKey algorithm to regenerate the keys. This can just an application that is not KERI but just the keeping.Manager .

This functionality was by design and has been there for several years now.

**Comment by SmithSamuelM** [comment]
Before KERI was Key history and the Key History was good so KERI could be created.

**Comment by SmithSamuelM** [comment]
One more thought.  As I outlined above, exporting private keys means simple adding an export function to the keeping.Manager in keripy so its easy. 

Whereas creating a standalone application to regenerate keys from the same salt, i.e. not exporting them. Is more work. 

The risk here is that once people take the easy path of exporting private keys, the tendency (very real) is to create fragile weak key mangement infrastructure because exporting private keys makes them shared secrets not private keys anymore and then build applications that depend on such shared secret infrastructure. This is how we got into the shared secret insecurity that is endemic of the web now.

**Comment by SmithSamuelM** [comment]
What many don't appreciate it is that 1password follows best practices for managing secrets so one can use it to manage salts for HDKeys.  1password uses slow hashing 0.25 seconds so that any reasonable length password used to unlock 1password can't be brute forced in any reasonable amount of time.  Likewise keripys keeping stretch algorithm uses slow hashing via the Argon2 algorithm that when set to the highest tier can take a long time to generate a key from a salt so guessing a salt to get a known key is not brute forceable.

So, basically, by not having any shared secrets, the attack vector is to get you to voluntarily unlock your 1 password or to intercept your salt when you copy it from 1password to unlock your Keri key store. But as soon as we enable exporting private keys, the attack vectors multiply without bounds.

**Comment by rodolfomiranda** [comment]
> 2\. Without a AID private key export feature via Signify, how could a KERIA account owner move their AIDs controlled by one KERIA Service to another provider or platform?  Could they request an export of their entire database, then import it to another KERIA instance, or decrypt its content with tools to extract key material needed?  Without these types of features, is an AID essentially locked in with the KERIA provider?

KERIA doen't have the private keys. Only the Signify client has them unencrypted, and it can potentially use them with different KERIA providers

**Comment by dhh1128** [comment]
We discussed on the community call on 10 Sep 2024, and agreed that we could achieve the goal without the suboptimal choice of exporting private keys. Fundamentally, we import the keys to KERI rather than creating them in KERI and exporting them.

@SmithSamuelM shared that in a habery's make() method, it is possible to pass a `secrecies` arg that allows the hab to be initialized with a pre-existing key. See roughly line 2190 and 2247 in the hab code (search for self.mgr.ingest). There are examples where this is used in tests/demos (see test_demos.py). What we would need to do is create an API that exposes this functionality so higher-level code could invoke it without direct access to the hab's make() method.

We also explored how to combine this with what Signify is doing. @pfeairheller pointed out that Signify already supports 3 AID creation methods (random, salt, HSM). We could add a 4th method that creates an AID from pre-existing keys, and then have Signify-TS just prompt for a privkey whenever it is needed.

I'm going to close this issue as resolved.

**Comment by iFergal** [comment]
Digging up this older issue...

@SmithSamuelM I'm also interested in applying this technique to bridge to a non KERI aware environment using delegation. In case one wants to revoke access to the non-transferable delegated identifier that we bridge to, e.g. ssh key, how does one accomplish that? (while keeping older signatures verifiable)

I presume we intend for it to be non-transferable so bare signatures (not interaction events) can be trusted. Or maybe this only works for ephemeral signatures like logins?

I want to do something such as signing github commits, and had thought about making it transferable. The rotation key could even be kept in the wallet itself (only signing keys in non-KERI aware environments). The problem is it would need to sign interaction events to be verifiable after the rotation.

Perhaps not possible without either having basic interaction event functionality in that environment, or using something like a blockchain for the timing.

**Comment by SmithSamuelM** [comment]
Not sure exactly why there would ever be a need to export and AID private keys in order to bridge to ecosystems that do not understand AIDs but do understand Ed25519. Why not just issue an ACDC that delegates a new non AID key pair for that bridged application. Since the bridged application can't make sense of AIDs and their key state only some party that can and wants to verify the provenance of some key pair will be able to. So the ACDC is the proveance evidence backing a bridged application key pair.  The bridged application creates the key pair and then exports the public key only back to the AID controller who then issues and ACDC entitling that new public key directily and hence indirectly through a signed challange and response the privat without exposing it.

**Comment by iFergal** [comment]
Thanks @SmithSamuelM - though I'm still curious how you can verify which signatures of the bridged application were made before and after the revocation of the ACDC, in case the bridged key is compromised/lost/etc.

Am I correct in, either:
- The bridged application needs to learn at least some KERI features, or,
- The signatures can only be verified in real time for acceptance into the platform (and potentially not verifiable afterwards), or,
- The application might already use something like a blockchain which can give that timeliness if it fits the use case.

**Comment by SmithSamuelM** [comment]
Frankly if you do not have a mechanism to anchor signatures to keystate, they signatures are only epherally verifiable against the current keystate and as soon as that key state changes the signatures become unverifiable.  

I you think about it, this is essentially how alternative identity systems work, an OIDC token is only valid for a short period of time, because the key state changes so rapidly, as soon as the keys rotate all the tokens are no longer verifiable, essentially automatically revoked by virtue of the rotation.   Same for an x509 certificate, key rotation is accomplished by issuing a new certificate, since there is no mechanism for revoking a cert, the cert must have a set expiration datetime.  

One could delegate a key pair with an ACDC that has an expiration date, essentially building a more secure version of x509 certs.  More secure because that CA in this case is KERI backed AID. But essentially unbound (unanchored signatures) really should only be used for ephemeral tasks like sending a request and verifying it dynamically against the current key state nor for any issuances that need to persist in time.

**Comment by iFergal** [comment]
Thanks @SmithSamuelM all makes sense! One final question for completeness - originally in this thread, it was suggested to delegate to a delegated identifier which was non-rotatable after inception. In this approach, can one "remove" the delegation if the key was compromised?

> Of course instead of using KERI one could anchor issuances to key state via a block chain. As you know KERI was meant to solve the same problem but be portable (i.e. not ledger locked as we used to say).

By this, I meant if it's already a use case where someone wants their witness-backed KERI identifier to delegate some authority to a key that will be used in a blockchain context. So the primary identifier remains (mostly) separate from the ledger.

**Comment by SmithSamuelM** [comment]
Currently delegated identifiers don't have an explicit mechanism for the delegator to "remove" the delegation. The mechanism for abandoning the identifier is to do a rotation by the delegatee (approved by the delegator) to an empty set of keys.
